src/Chainweb/Pact/PactService/Checkpointer.hs

```diff
@@ -149,379 +86,89 @@ exitOnRewindLimitExceeded = handle $ \case
-            <> "; target hash: " <> blockHashToText parentHash
-
-        failOnTooLowRequestedHeight parent rewindLimit lastHeader
-        playFork lastHeader
-
-  where
-    parentHeight = _blockHeight parent
-    parentHash = _blockHash parent
-
-    failNonGenesisOnEmptyDb = error "impossible: playing non-genesis block to empty DB"
-
-    playFork :: BlockHeader -> PactServiceM logger tbl ()
-    playFork lastHeader = do
-        bhdb <- asks _psBlockHeaderDb
-        commonAncestor <- liftIO $ forkEntry bhdb lastHeader parent
-        let ancestorHeight = _blockHeight commonAncestor
-
-        if commonAncestor == parent
-          then
-            -- If no blocks got replayed the checkpointer isn't restored via
-            -- 'fastForward'. So we do an empty 'withCheckPointerWithoutRewind'.
-            withCheckpointerWithoutRewind (Just $ ParentHeader commonAncestor) "rewindTo" $ \_ ->
-                return $! Save commonAncestor ()
-          else do
-            logInfo $ "rewindTo.playFork"
-                <> ": checkpointer is at height: " <> sshow (_blockHeight lastHeader)
-                <> ", target height: " <> sshow (_blockHeight parent)
-                <> ", common ancestor height " <> sshow ancestorHeight
-
-            logger <- view psLogger
-            -- 'getBranchIncreasing' expects an 'IO' callback because it maintains an 'TreeDB'
-            -- iterator. 'withPactState' allows us to call pact service actions
-            -- from the callback.
-            c <- withPactState $ \runPact ->
-                getBranchIncreasing bhdb parent (int ancestorHeight) $ \newBlocks -> do
-                    -- This stream is guaranteed to at least contain @e@.
-                    (h, s) <- fromJuste <$> S.uncons newBlocks
-                    heightRef <- newIORef (_blockHeight commonAncestor)
-                    withAsync (heightProgress (_blockHeight commonAncestor) heightRef (logInfo_ logger)) $ \_ ->
-                      s
-                          & S.scanM
-                              (\ !p !c -> runPact (fastForward (ParentHeader p, c)) >> writeIORef heightRef (_blockHeight c) >> return c)
-                              (return h) -- initial parent
-                              return
-                          & S.length_
-            logInfo $ "rewindTo.playFork: replayed " <> sshow c <> " blocks"
-
--- | INTERNAL UTILITY FUNCTION. DON'T EXPORT FROM THIS MODULE.
---
--- Fast forward a block within a 'rewindTo' loop.
---
-fastForward
-    :: forall logger tbl
-    . (HasCallStack, CanReadablePayloadCas tbl, Logger logger)
-    => (ParentHeader, BlockHeader)
-    -> PactServiceM logger tbl ()
-fastForward (target, block) = do
-    -- This does a restore, i.e. it rewinds the checkpointer back in
-    -- history, if needed.
-    withCheckpointerWithoutRewind (Just target) "fastForward" $ \pdbenv -> do
-        payloadDb <- asks _psPdb
-        payload <- liftIO $ tableLookup payloadDb bpHash >>= \case
```

Comment:
We need a test in master that covers fastForward scenario and then check that the test works fine in this branch — the only way to check we don't break anything here.

Reply:
I basically inlined `fastForward` and its friends into its callers.